documentation/projects/proposals/proxy_frontend_api_requests/20241202-implementation_plan_proxy_frontend_api_requests_ip.md

```diff
@@ -0,0 +1,396 @@
+
+Currently, client requests[^1] from the frontend to the API are sent without
+authentication. As a result, strict API rate limiting on unauthenticated
+requests would inadvertently throttle legitimate traffic from openverse.org
+users. By proxying all client requests[^1] through the Nuxt server, we can
+authenticate them and thus distinguish them from the anonymous traffic in the
+API.
+
+This implementation plan outlines how we will route all frontend traffic to the
+API through Nuxt server routes — effectively “authenticating” these requests at
+the API level — and configure Cloudflare rate-limiting to protect these routes
+from excessive automated traffic that is against Openverse Terms of Service.
+
+While this plan does not include lowering unauthenticated rate limits in the
+API, it establishes the groundwork for it. This approach will ensure better
+stability and fairness of the API service, while also preserving the normal user
+experience on the frontend.
+
+## Expected Outcomes
+
+1. **Distinguishable Frontend Requests in the API**: Frontend requests are
+   authenticated at the API level and no longer constrained by unauthenticated
+   rate limits.
+
+2. **Protected Nuxt Server Routes and Fair Access**: Nuxt server routes are
+   protected from automated abuse and suspicious traffic by Cloudflare
+   rate-limiting, ensuring that legitimate users (including those in shared IP
+   environments, such as schools and libraries) maintain reliable access without
+   undue friction.
+
+## Request Flow
+
+From the Nuxt client perspective, the request flow will not change much: instead
+of going directly to the API, requests will go to the Nuxt server routes
+(`/api/**`). The main difference is that, when rate-limited, the client will
+receive a Cloudflare challenge response that will be handled by
+[Nuxt Turnstile Module](https://github.com/nuxt-modules/turnstile).
+
+The Nuxt server routes will authenticate the requests by adding the API token,
+and sign requests with HMAC to allow Cloudflare to apply rate-limiting on a
+per-user basis, even when multiple users share the same IP (e.g., in NAT
+environments).
+
+```{mermaid}
+sequenceDiagram
+    participant O as Client
+    participant C as Cloudflare
+    participant NS as Nuxt Server
+    participant A as API
+    O ->> C: HMAC-signed request
+    Note right of C: "Check the the rate limits based on cookie&IP"
+    alt successful
+        C ->> NS: request
+        NS ->> A: authenticated request
+        A ->> NS: response
+        NS ->> O: response
+    else rate-limited
+        C ->> O: turnstile widget
+        O ->> C: Cloudflare challenge cookie
+        C ->> NS: request
+        NS ->> A: authenticated request
+        A ->> NS: response
+        NS ->> O: response
+    end
+```
+
+## Cloudflare Rate Limiting
+
+Originally, I thought of using Cloudflare managed challenges to protect the Nuxt
+server routes from abuse. However, this would have required issuing a challenge
```

Comment:
A clarification here: "to protect the Nuxt server routes from abuse".

For the solution to work, it's necessary for _all_ routes to be behind the same Cloudflare rate limit rules, not just the server routes. That way, if someone gets rate limited on the server routes, the frontend can reload the page and the user will actually get the managed challenge page for clearing the rate limiting (or the explanation page if they are going too fast and need to actually be limited).

If only the server routes are limited by the new configuration, then there's no way to present the user with a challenge that will clear that rate limit, and the user would be stuck.

I mention in another comment that I don't think this would require issuing a challenge to every new user and would not meaningfully increase friction in the vast majority of even extreme cases. Regardless, it is necessary for all routes to have the same rate limit rules for this Cloudflare-rate-limit strategy to work with the safe guard of users being able to clear the rate limit as needed via the challenge.

Reply:
Gotcha! I was not connecting the dots on the turnstile handling the flow of presenting the challenge for the subset of rate limited routes.